### PR TITLE
Add error-handling slides.

### PR DESCRIPTION
**Pull request by jonathanpallant**
Adds slides on Error Handling (`./src/error-handling.md`).

Talks about `&'static str`, `String`, `enum Error`, `std::error::Error`, `thiserror` and `anyhow`.

### CODE REVIEW

File: training-slides/src/error-handling.md [hunk 1]
```diff
@@ -1 +1,178 @@
...
+## There are no exceptions
+
+Rust has two ways of indicating errors:
+
+* Returning a value
+* Panicking
+
+## Returning Values
+
+When a function can *either* be __Ok__ or can return an __Error__, we use `Result<T, E>`:
```
Comment (amanjeev) [inline training-slides/src/error-handling.md:12]: small nit: I would rephrase this because both `Ok` and `Error` are returned right?  "can either be Ok" feels like might not return?

File: training-slides/src/error-handling.md [hunk 2]
```diff
@@ -1 +1,178 @@
...
+}
+```
+
+*Why might this function fail?*
+
+## What kind of Error?
+
+You can put anything in for the `E` in `Result<T, E>`:
+
+* A `&'static str`: `"Number overflow"`
```
Comment (amanjeev) [inline training-slides/src/error-handling.md:28]: I would like to rephrase this too - 

- A value of type `&'static str`; e.g. "Number overflow"
- A value of type `String`; e.g. `format!("Number {} too large", sum)`
- A value of type `enum`; e.g. `Error::NumberOverflow(sum)`
- etc.

I am trying to avoid `:` because it confuses my brain lol syntax vs. english.

File: training-slides/src/error-handling.md [hunk 3]
```diff
@@ -1 +1,178 @@
...
+fn some_function(works: bool) -> Result<u32, &'static str> {
+    if works {
+        Ok(42)
+    } else {
+        Err("I'm not working today")
+    }
+}
+```
+
+## ? also called '.into()' for you
```
Comment (amanjeev) [inline training-slides/src/error-handling.md:53]: `?` also called `.into()` for you

I wonder if syntax highlighting in headers also works.  Trying to keep it consistent.

File: training-slides/src/error-handling.md [hunk 4]
```diff
@@ -1 +1,178 @@
...
+* It's expressive
+* But you can't change the string to include some specific value
+* And your program can't tell what *kind* of error it was
+
+## Using Strings
+
+Setting `E` to be `String` lets you use heap-allocated strings:
+
+* It's expressive
+* You can include some specific values in the String
```
Comment (amanjeev) [inline training-slides/src/error-handling.md:84]: `String`

(syntax for String)

File: training-slides/src/error-handling.md [hunk 5]
```diff
@@ -1 +1,178 @@
...
+* But it costs you a heap allocation to store the bytes for the String
```
Comment (amanjeev) [inline training-slides/src/error-handling.md:85]: `String`

(syntax for String)

File: training-slides/src/error-handling.md [hunk 6]
```diff
@@ -1 +1,178 @@
...
+## Something universal
+
+Exhaustively listing all the ways your dependencies can fail is hard.
+
+One solution:
+
+```rust [1-5|1|2|3] should_panic
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let f = std::fs::File::open("hello.txt")?; // IO Error
+    let s = std::str::from_utf8(&[0xFF, 0x65])?; // Unicode conversion error
```
Comment (amanjeev) [inline training-slides/src/error-handling.md:163]: Do you think using `_f` and `_s` here would help because the pesky warnings will go away and the error you want to show will be shown.

File: training-slides/src/error-handling.md [hunk 7]
```diff
@@ -1 +1,178 @@
...
+    Ok(())
+}
+```
+
+## Anyhow
+
+The [`anyhow`](https://crates.io/crates/anyhow) crate gives you a nicer type:
+
+```rust [1-5|1] ignore
```
Comment (amanjeev) [inline training-slides/src/error-handling.md:172]: Is there no way to make this run?  because no way to import/use anyhow in code examples? :(

**Review by amanjeev — CHANGES_REQUESTED** [review]
I have some minor nits.

File: training-slides/src/error-handling.md (hunk 4 above)
Comment (jonathanpallant) [reply training-slides/src/error-handling.md:84]: You can offer these as suggestions, using the [+-] icon, or writing a markdown comment like:

````markdown
```suggestion
* You can include some specific values in the `String`
```
````

File: training-slides/src/error-handling.md (hunk 2 above)
Comment (jonathanpallant) [reply training-slides/src/error-handling.md:28]: I replaced this with a block of Rust code which hopefully is easier to grok.

File: training-slides/src/error-handling.md (hunk 1 above)
Comment (jonathanpallant) [reply training-slides/src/error-handling.md:12]: Fixed

**Comment by jonathanpallant** [comment]
Updated based on changes.

File: training-slides/src/error-handling.md (hunk 6 above)
Comment (amanjeev) [reply training-slides/src/error-handling.md]: Yes when you click 'run" button it shows the unused variable warning.

File: training-slides/src/error-handling.md (hunk 7 above)
Comment (amanjeev) [reply training-slides/src/error-handling.md:190]: Yea lets not do that it is not nice indeed.

**Review by amanjeev — APPROVED** [review]
Thank you!
